EliminateLeftJoin rule added

ronaksingh27 opened the pull request.

ronaksingh27 committed:
EliminateLeftJoin rule added

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=132518) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=132518) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=132518) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/132518)
<!-- Reviewable:end -->

ronaksingh27 commented on pkg/sql/opt/norm/rules/join.opt:
```diff
@@ -369,6 +369,16 @@ $left
+    $right:* & (HasZeroRows $right)
+)
+=>
+$left
```
Thank you, @DrewKimball  I will certainly take a closer look and work on creating a more accurate solution.

ronaksingh27 committed:
feat: add Passthrough() and MakeNullProjections() logic for optimizer

- Added Passthrough() and MakeNullProjections() functions.
- Modified the logic for Optimizer.
- Resolved extra line at line 386.

DrewKimball reviewed (commented):
Sorry for the delay. This is looking good to me! I just had a few nit-picky suggestions. At this point, you'll want to start thinking about how to exercise this logic with some optimizer tests. You can add a new section for your rule to `pkg/sql/opt/norm/testdata/rules/join` just like the tests for the `ProjectInnerJoinValues` rule here:
https://github.com/cockroachdb/cockroach/blob/1e5b3c212b45419c960038718c48a5dd75a111a0/pkg/sql/opt/norm/testdata/rules/join#L3913-L3915

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132518#-:-O9mil9E-081tWHTUavj:bh4evv1)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @ronaksingh27 and @rytaft)

___
*[`pkg/sql/opt/norm/general_funcs.go` line 522 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/132518#-O9mi1ek9wIzDtRVAWDL:-O9mi1ek9wIzDtRVAWDM:b-8naijt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee20b254ad1e76b8fc73769eb4c802454d7a8229/pkg/sql/opt/norm/general_funcs.go#L522)):*
> ```Go
> 
>     // For each column in the right input, create a NULL projection.
>     for _, col := range rightCols.Ordered() {
> ```

[nit] Using `Ordered` will force us to allocate a new slice just to iterate through the set; you can use this pattern to avoid allocations:
```
for col, ok := rightCols.Next(0); ok; col, ok = src.rightCols.Next(col + 1) {
  ...
}
```
___
*[`pkg/sql/opt/norm/general_funcs.go` line 531 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/132518#-O9mhAE45YFHulwess3Z:-O9mhAE45YFHulwess3_:b-qw0224) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee20b254ad1e76b8fc73769eb4c802454d7a8229/pkg/sql/opt/norm/general_funcs.go#L531)):*
> ```Go
> 
> // PassthroughCols returns the set of columns from the left input to be passed through.
> func (c *CustomFuncs) PassthroughCols(left memo.RelExpr) opt.ColSet {
> ```

[nit] Rather than adding a whole new custom function just for this specific usage, let's just use the existing `OutputCols` custom func.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
Hi @ronaksingh27, do you think you'll have a chance to pick this back up?

ronaksingh27 commented:
@DrewKimball yeah sure give me some time I may need to go through everything again , apologies for delay
